hw/dv/dpi/usbdpi/usbdpi.c

```diff
@@ -102,6 +102,77 @@ static void bus_reset(usbdpi_ctx_t *ctx) {
 static void usbdpi_data_callback(void *ctx_v, usbmon_data_type_t type,
                                  uint8_t d);
 
+// Sample the bus signals
+static bool bus_sample(usbdpi_ctx_t *ctx, unsigned *pdp, unsigned *pdn,
+                       uint32_t d2p) {
+  assert(ctx);
+
+  // Ascertain the state of the D+/D- signals from the device
+  unsigned dp, dn;
+  if (d2p & D2P_TX_USE_D_SE0) {
+    // Single-ended mode uses D and SE0
+    if (d2p & D2P_D_EN) {
+      if (d2p & D2P_DNPU) {
+        // Pullup says swap i.e. D is inverted
+        dp = (d2p & D2P_SE0) ? 0 : ((d2p & D2P_D) ? 0 : 1);
+        dn = (d2p & D2P_SE0) ? 0 : ((d2p & D2P_D) ? 1 : 0);
+      } else {
+        dp = (d2p & D2P_SE0) ? 0 : ((d2p & D2P_D) ? 1 : 0);
+        dn = (d2p & D2P_SE0) ? 0 : ((d2p & D2P_D) ? 0 : 1);
+      }
+    } else {
+      dp = (d2p & D2P_PU) ? 1 : 0;
+      dn = 0;
+    }
+  } else {
+    // Normal D+/D- mode
+    if (d2p & D2P_DNPU) {
+      // Assertion of DN pullup suggests DP and DN are swapped
+      dp = ((d2p & D2P_DN_EN) && (d2p & D2P_DN)) ||
+           (!(d2p & D2P_DN_EN) && (d2p & D2P_DNPU));
+      dn = (d2p & D2P_DP_EN) && (d2p & D2P_DP);
+    } else {
+      // No DN pullup so normal orientation
+      dp = ((d2p & D2P_DP_EN) && (d2p & D2P_DP)) ||
+           (!(d2p & D2P_DP_EN) && (d2p & D2P_DPPU));
+      dn = (d2p & D2P_DN_EN) && (d2p & D2P_DN);
+    }
+  }
+
+  // Return state of DP/DN signals
+  if (pdp)
+    *pdp = dp;
+  if (pdn)
+    *pdn = dn;
+
+  // If the bus is idle but we've detected a non-idle state, adjust the sampling
+  // phase; this synchronization mimics that of the usbdev logic, although we
+  // could potentially do something more sophisticated over the 6 or 8 bit
+  // intervals of the known SYNC (KJKJKJKK) sequence that starts each packet.
+  if (ctx->state == ST_IDLE) {
```

Comment:
This isn't quite as sophisticated as usbdev, which adapts the sampling point on every transition. I'm not sure we can get away without it, if we're going to test the limits.

At the 48 MHz sampling clock and for 64 raw bytes without bitstuffing (64 bytes * 8 bits / byte * 4 clocks / bit), we end up sampling 1 clock early for the next bit with only a 24 kHz offset (0.049%, 488 ppm).